src/execution/operator/persistent/physical_insert.cpp

```diff
@@ -343,9 +363,85 @@ idx_t PhysicalInsert::OnConflictHandling(TableCatalogEntry &table, ExecutionCont
+	auto &local_storage = LocalStorage::Get(context.client, data_table.db);
+
+	// We either want to do nothing, or perform an update when conflicts arise
+	ConflictInfo conflict_info(conflict_target);
+	ConflictManager conflict_manager(VerifyExistenceType::APPEND, lstate.insert_chunk.size(), &conflict_info);
+	DataTable::VerifyUniqueIndexes(local_storage.GetIndexes(data_table), context.client, lstate.insert_chunk,
+	                               &conflict_manager);
+	conflict_manager.Finalize();
+	if (conflict_manager.ConflictCount() == 0) {
+		// No conflicts found, 0 updates performed
+		return 0;
+	}
+	auto &conflicts = conflict_manager.Conflicts();
+	auto &row_ids = conflict_manager.RowIds();
+
+	DataChunk conflict_chunk; // contains only the conflicting values
+	DataChunk scan_chunk;     // contains the original values, that caused the conflict
+	DataChunk combined_chunk; // contains conflict_chunk + scan_chunk (wide)
+
+	// Filter out everything but the conflicting rows
+	conflict_chunk.Initialize(context.client, lstate.insert_chunk.GetTypes());
+	conflict_chunk.Reference(lstate.insert_chunk);
+	conflict_chunk.Slice(conflicts.Selection(), conflicts.Count());
+	conflict_chunk.SetCardinality(conflicts.Count());
+
+	// Holds the pins for the fetched rows
+	unique_ptr<ColumnFetchState> fetch_state;
+	if (!types_to_fetch.empty()) {
+		D_ASSERT(scan_chunk.size() == 0);
+		// When these values are required for the conditions or the SET expressions,
+		// then we scan the existing table for the conflicting tuples, using the rowids
+		scan_chunk.Initialize(context.client, types_to_fetch);
+		fetch_state = make_uniq<ColumnFetchState>();
+		local_storage.FetchChunk(data_table, row_ids, conflicts.Count(), columns_to_fetch, scan_chunk, *fetch_state);
+	}
+
+	// Splice the Input chunk and the fetched chunk together
+	CombineExistingAndInsertTuples(combined_chunk, scan_chunk, conflict_chunk, context.client);
+
+	if (on_conflict_condition) {
+		DataChunk conflict_condition_result;
+		CheckOnConflictCondition(context, combined_chunk, on_conflict_condition, conflict_condition_result);
+		bool conditions_met = AllConflictsMeetCondition(conflict_condition_result);
+		if (!conditions_met) {
+			// Filter out the tuples that did pass the filter, then run the verify again
+			ManagedSelection sel(combined_chunk.size());
+			auto data = FlatVector::GetData<bool>(conflict_condition_result.data[0]);
+			for (idx_t i = 0; i < combined_chunk.size(); i++) {
+				if (!data[i]) {
+					// Only populate the selection vector with the tuples that did not meet the condition
+					sel.Append(i);
+				}
+			}
+			combined_chunk.Slice(sel.Selection(), sel.Count());
+			row_ids.Slice(sel.Selection(), sel.Count());
+			DataTable::VerifyUniqueIndexes(local_storage.GetIndexes(data_table), context.client, lstate.insert_chunk,
+			                               nullptr);
+			throw InternalException("The previous operation was expected to throw but didn't");
```

Comment:
I am curious as to why we need to run this again. Do you know if this is only relevant for debugging purposes? Especially because it throws an internal exception, triggering this with user input should not be possible. If it is for debugging only, maybe we should add `#ifdef DEBUG`? Same for the code in `HandleGlobalConflicts`.

Reply:
We've found a conflict, we checked the condition that it should meet to be picked up by the DO UPDATE|NOTHING and it doesnt meet it.

So this should throw an error.
Because we would only get here if we found a conflict, we can safely rerun the check to get the error.

The internal exception should never happen because that code should be unreachable, I'd prefer to verify that that is the case on release as well